subsys/emds/emds.c

```diff
@@ -202,7 +149,43 @@ int emds_store(void)
 	store_key = irq_lock();
 
 	/* Start the emergency data storage process. */
-	k_sem_give(&emds_sem);
+	LOG_DBG("Emergency Data Storeage released");
+
+	STRUCT_SECTION_FOREACH(emds_entry, ch) {
+		ssize_t len = emds_flash_write(&emds_flash,
+					       ch->id, ch->data, ch->len);
+		if (len < 0) {
+			LOG_ERR("Write static entry: (%d) error (%d)",
+				ch->id, len);
+		} else if (len != ch->len) {
+			LOG_ERR("Write static entry: (%d) failed (%d:%d)",
+				ch->id, ch->len, len);
+		}
+	}
+
+	struct emds_dynamic_entry *ch;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+		ssize_t len = emds_flash_write(&emds_flash,
+					       ch->entry.id, ch->entry.data, ch->entry.len);
+		if (len < 0) {
+			LOG_ERR("Write dynamic entry: (%d) error (%d).",
+				ch->entry.id, len);
+		}
+		if (len != ch->entry.len) {
+			LOG_ERR("Write dynamic entry: (%d) failed (%d:%d).",
+				ch->entry.id, ch->entry.len, len);
+		}
+	}
+
+	emds_ready = false;
+
+	/* Unlock all interrupts */
+	irq_unlock(store_key);
```

Comment:
Since this flow is refactored, can we then move this to the end of this function? I think it will make it streamlined.

Reply:
This also keep the functionality as it was before. This keep the irq lock just as long as it has to.